File: Day7.py
inputString = open("Day7Input.txt", "r").read().splitlines()
bags, foundBags, containCount = {}, set(), 0

def readInput():
    for line in inputString:
        bags[line.split(" contain ")[0]] = line.split(" contain ")[1]

def findBags(searchString):
    containers = {keys for keys, values in bags.items() if searchString in values}
    if len(containers) > 0:
        foundBags.update(containers)
        for container in containers:
            findBags(container[:-5])

def countContains(searchString, multiplier):
    global containCount
    if bags[searchString] != "no other bags.":
        for subBag in bags[searchString].split(", "):
            containCount += multiplier * int(subBag.split(" ")[0])
            countContains(subBag.split(" bag")[0][2:].strip() + " bags", multiplier * int(subBag.split(" ")[0]))

readInput()
findBags("shiny gold")
print(len(foundBags))
countContains("shiny gold bags", 1)
print(containCount)

File: CSharp/Day7.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Linq;

namespace CSharp
{
    class Day7
    {
        public static void getSolutions()
        {
            string[] inputData = File.ReadAllLines("../../../Day7Input.txt");
            Dictionary<string, string> inputDict = readInputData(inputData);
            RunPart1(inputDict);
            RunPart2(inputDict);
        }

        static Dictionary<string, string> readInputData(string[] inputData)
        {
            Dictionary<string, string> outputDict = new();
            foreach (string line in inputData)
            {
                outputDict[line.Split(" contain ")[0].Trim()] = line.Split(" contain ")[1].Trim();
            }
            return outputDict;
        }

        static void RunPart1(Dictionary<string, string> inputDict)
        {
            HashSet<string> bags = new();
            lookupBag("shiny gold", bags, inputDict);
            Console.WriteLine(bags.Count);
        }

        static void lookupBag(string searchString, HashSet<string> bags, Dictionary<string, string> inputDict)
        {
            IEnumerable<string> result = inputDict.Where(p => p.Value.Contains(searchString)).Select(p => p.Key);
            foreach (string newBag in result)
            {
                bags.Add(newBag[0..^5]);
                lookupBag(newBag[0..^5].Trim(), bags, inputDict);
            }
        }

        static void RunPart2(Dictionary<string, string> inputDict)
        {
            Console.WriteLine(countContains("shiny gold bags", 1, inputDict));
        }

        static long countContains(string keyString, long multiplier, Dictionary<string, string> inputDict)
        {
            long containCount = 0;
            if (inputDict[keyString] != "no other bags.")
            {
                foreach (string subBag in inputDict[keyString].Split(", "))
                {
                    containCount += long.Parse(subBag.Split(" ")[0]) * multiplier;
                    containCount += countContains(subBag[2..].Split("bag")[0].Trim() + " bags", long.Parse(subBag.Split(" ")[0]) * multiplier, inputDict);
                }
            }
            return containCount;
        }
    }
}
